File: code/605.can-place-flowers.php
<?php
/*
 * @lc app=leetcode id=605 lang=php
 *
 * [605] Can Place Flowers
 */

// @lc code=start
class Solution {

    /**
     * @param Integer[] $flowerbed
     * @param Integer $n
     * @return Boolean
     */
    function canPlaceFlowers($flowerbed, $n) {
        // e.g.
        // $count = 0;
        // $flowerbedLength = count($flowerbed);
        // for ($i = 0; $i <= $flowerbedLength; $i++) {
        //     if ($flowerbed[$i] === 0
        //     && $i < $flowerbedLength) { // 最後の要素が0の場合にfalseに行くように
        //         $count++;
        //         if ($i === 0) $count++;
        //         if ($i === $flowerbedLength - 1) $count++;
        //     } else {
        //         $n -= intdiv($count - 1, 2);
        //         if ($n <= 0) return true;
        //         $count = 0;
        //     }
        // }
        // return false;

        // e.g.
        array_unshift($flowerbed, 0);
        $flowerbed[] = 0;
        for ($i = 1; $i < count($flowerbed) - 1; $i++) {
            if ($flowerbed[$i - 1] + $flowerbed[$i] + $flowerbed[$i + 1] === 0) {
                $n--;
                $i++; // 今の要素に1が入るため1要素分飛ばしてみる
            }
            if ($n <= 0) return true;
        }
        return $n <= 0;
    }
}
// @lc code=end

File: code/605.can-place-flowers.go
package code

/*
 * @lc app=leetcode id=605 lang=golang
 *
 * [605] Can Place Flowers
 */

// @lc code=start
func canPlaceFlowers(flowerbed []int, n int) bool {
	// 先頭と末尾に0を追加
	flowerbed = append([]int{0}, flowerbed...)
	flowerbed = append(flowerbed, 0)

	for i, v := range flowerbed {
		if i == 0 || i == len(flowerbed)-1 {
			continue
		}

		if flowerbed[i-1] == 0 && v == 0 && flowerbed[i+1] == 0 {
			flowerbed[i] = 1
			n--
			// 次の要素は確定で0なので飛ばす
			i++
		}

		if n <= 0 {
			return true
		}
	}

	return false
}

// @lc code=end
